src/controllers/user.js

```diff
@@ -74,4 +77,80 @@ module.exports = {
       res.status(422).json({ message: err.message ?? err });
     }
   },
+  rateUser: async (req, res) => {
+    const { userid } = req.params;
+    try {
+      if (String(new ObjectId(userid)) !== userid.toString())
+        throw new Error('Requested user ID is not valid!');
+      const user = await UserModel.findById(userid).populate(
+        'rating',
+        'raters'
+      );
+      if (!user) throw new Error("The user with the specified ID wasn't found");
+      if (!user.rating) {
+        const newRatingData = {
+          user: userid,
+          raters: [
+            {
+              raterId: req.user._id,
+              rating: req.body.rating,
+            },
+          ],
+        };
+        const newRating = await RatingModel.create(newRatingData);
+        user.rating = newRating.id;
+        await user.save();
+        res.status(201).json(newRating);
+      } else {
+        const newRater = {
+          raterId: req.user._id,
+          rating: req.body.rating,
+        };
+        const rating = await RatingModel.findById(user.rating);
+        /* eslint no-restricted-syntax: ["off"] */
+        /* eslint prefer-const: "off" */
+        for (let rater of user.rating.raters) {
+          if (rater.raterId.toString() === req.user._id.toString())
+            throw new Error('You had already rated this user');
+        }
+        rating.raters.push(newRater);
+        await rating.save();
+        res.status(201).json(rating);
+      }
+    } catch (err) {
+      res.status(422).json({ message: err.message ?? err });
+    }
+  },
+  updateUserRating: async (req, res) => {
+    const { userid } = req.params;
+    try {
+      if (String(new ObjectId(userid)) !== userid.toString())
+        throw new Error('Requested user ID is not valid!');
+      const user = await UserModel.findById(userid).populate(
+        'rating',
+        'raters'
+      );
+      if (!user) throw new Error("The user with the specified ID wasn't found");
+      const rating = await RatingModel.findById(user.rating);
+      if (!rating) throw new Error('The user does not have any rating');
+      const userRaters = user.rating.raters;
+      let raterExists = false;
+      let indexOfRater = 0;
+      /* eslint no-restricted-syntax: ["off"] */
+      /* eslint prefer-const: "off" */
+      for (let rater of userRaters) {
+        if (rater.raterId.toString() === req.user._id.toString()) {
+          raterExists = true;
+          rating.raters[indexOfRater].rating = req.body.rating;
+        }
+        indexOfRater += 1;
+      }
```

Comment:
Just a suggestion for this loop. I think you can check if the current authenticated user has rated that user before like this  `const currentUsersRate = userRaters.find((rater)=>rater.raterId.toString() === req.user._id.toString())` because in the `rateUser` you are making sure that an authenticated user rate a user only once. So, I think you can get rid of for loop and `indexOfRater`

Reply:
I agree with Mehmet, but I think you need the rater index so you can rewrite this as:
```js
const indexOfRater = userRaters.findIndex(rater => rater.raterId.toString() === req.user._id.toString();
if (indexOfRater > -1) rating.raters[indexOfRater].rating = req.body.rating;
```